### PR TITLE
Loki Derived Fields: Refactor legacy form components and add validation

### LINKED ISSUE #52598: loki: derived fields with the same name do not work correctly
if you create two derived-fields with the same `name` in the loki datasource config, it will behave incorrectly in the logs-view, in some cases it will create two links and in some cases no links.

how to reproduce:
1. `make devenv sources=loki`
2. create a loki datasource connecting to localhost:3100
    - add two derived fields with the same name
        - regex1: counter=(\d+)
        - regex2: counter":"(\d+)"
    - make sure they have an URL specified
3. go to explore-mode, run this loki query: `{source="data"}`, and start opening the log-details. in some cases, you get two links on the derived-field, and in some cases you get no links on the derived-field.

NOTE: i'm not sure whether this should even be allowed. one solution is to just do not allow two variables with the same name. though, this can be useful, like in this example, you have a mix of `logfmt` and `json` log-lines, and you want to extract traceids from both. the problem is, if we allow two derived-fields with the same name, how does it work when your log-line contains both formats? need to think about this when solving it.

### PR DESCRIPTION
**Pull request by matyax**
Follow up of https://github.com/grafana/grafana/pull/67220  . Here, we refactor the form to use non-legacy components, and we display the name error in a more nicely way, as part of https://github.com/grafana/grafana/issues/65294

Also fixed a bug where the data source and internal bug would not be saved if the user did not select a data source.

https://user-images.githubusercontent.com/1069378/236854850-6b23e2c2-191e-4b09-9eda-055b1287d7e2.mov

It should be expected to work on follow up PRs related with https://github.com/grafana/grafana/issues/65294 as common components become available.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/52598

Part of https://github.com/grafana/grafana/issues/65294

**Special notes for your reviewer:**

Before:
Before

After:
After

Please note that this form might become stacked in one column instead of two. The main purpose of this PR is to refactor legacy components and add missing validation.

### CODE REVIEW

**Comment by matyax** [comment]
FYI @catherineymgui

**Comment by matyax** [comment]
FYI @catherineymgui @josiahg

File: public/app/plugins/datasource/loki/configuration/DerivedField.tsx [hunk 1]
```diff
@@ -60,101 +65,108 @@ export const DerivedField = (props: Props) => {
...
-            }
-            setShowInternalLink(!showInternalLink);
-          }}
-        />
+      <div className="gf-form">
+        <Field label="Internal link" className={styles.internalLink}>
+          <Switch
+            value={showInternalLink}
+            onChange={(e: ChangeEvent<HTMLInputElement>) => {
+              const checked = e.currentTarget.checked;
```
Comment (gabor) [inline public/app/plugins/datasource/loki/configuration/DerivedField.tsx:133]: nitpick: if you want to you can write it this way too, but it's as you prefer.

```suggestion
              const { checked } = e.currentTarget;
```

**Review by gabor — CHANGES_REQUESTED** [review]
code is fine 👍 ,
there's one thing that did not work for me. the regular-expression does not get saved for me.

1. create a derived field, assign it a name, enter something in the regex-field
2. press [save&test]
3. reload the page
4. the field's name is correct, but the regex-field is empty

**Review by gabor — APPROVED** [review]
LGTM 👍

**Pull request merged**